clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java

```diff
@@ -644,6 +645,39 @@ public void shouldThrowIllegalStateExceptionOnAppendWhenAborted() throws Excepti
         return values;
     }
 
+    @Test
+    public void testBuffersDereferencedOnClose() {
+        Runtime runtime = Runtime.getRuntime();
+        int payloadLen = 1024 * 1024;
+        ByteBuffer buffer = ByteBuffer.allocate(payloadLen * 2);
+        byte[] key = new byte[0];
+        byte[] value = new byte[payloadLen];
+        new Random().nextBytes(value); // Use random payload so that compressed buffer is large
+        List<MemoryRecordsBuilder> builders = new ArrayList<>(100);
+        long startMem = 0;
+        long memUsed = 0;
+        int iterations =  0;
+        while (iterations++ < 100) {
+            buffer.rewind();
+            MemoryRecordsBuilder builder = new MemoryRecordsBuilder(buffer, RecordBatch.MAGIC_VALUE_V2, compressionType,
+                    TimestampType.CREATE_TIME, 0L, 0L, RecordBatch.NO_PRODUCER_ID,
+                    RecordBatch.NO_PRODUCER_EPOCH, RecordBatch.NO_SEQUENCE, false, false,
+                    RecordBatch.NO_PARTITION_LEADER_EPOCH, 0);
+            builder.append(1L, new byte[0], value);
+            builder.build();
+            builders.add(builder);
+
+            System.gc();
+            memUsed = runtime.totalMemory() - runtime.freeMemory() - startMem;
+            // Ignore memory usage during initialization
+            if (iterations == 2)
+                startMem = memUsed;
+            else if (iterations > 2 && memUsed < (iterations - 2) * 1024)
+                break;
+        }
+        assertTrue("Memory usage too high: " + memUsed, iterations < 100);
```

Comment:
Are we confident that this test is stable? An alternative is to verify that the fields are unset after close. Not as effective, but faster and not flaky.

Reply:
@ijuma Thanks for the review. The test retries for 100 iterations and I haven't seen it fail. If we see any failures, then I can check for the fields instead.

Reply:
And the test fails without the change?

Reply:
@ijuma Test fails consistently for me for LZ4 and GZIP without the change. Test doesn't fail with SNAPPY since memory growth is slower, but since the changes in the PR are verified with LZ4 and GZIP, I thought that was ok.

Reply:
Sounds good.

Reply:
Based on my analysis of the issue, this fix should definitely address the LZ4 and looks good to me.

Does it also fix the same issue in GZIP?  My testing ran out of memory using GZIP as well, but I don't see any gzip classes included.

Reply:
@Lothsahn Thanks for the review. With the change to set `MemoryRecordsBuilder.appendStream` to `null` when the builder is closed, this should fix the issue with GZIP as well.

Reply:
@rajinisivaram Ah.  I missed that section.  Good fix!

LGTM